text/tls-and-alpn.rst

```diff
@@ -0,0 +1,380 @@
+accept password as an environment variable ``EDGEDB_TLS_PRIVATE_KEY_PASSWORD``.
+
+If specified, ``--tls-ciphers`` will be used to call
+``ssl.SSLContext.set_ciphers()``.
+
+We'll discuss later in Core EdgeDB Development for the case of omitting ``tls``
+parameters.
+
+
+CLI Tooling
+-----------
+
+For the ``edgedb`` CLI, the goal is to make TLS in EdgeDB easy to use. There
+are 2 major scenarios:
+
+1. Using EdgeDB with TLS in development process.
+2. Deploy EdgeDB with TLS for production environment.
+
+As we'd like to enable TLS by default, a certificate is always required before
+starting an EdgeDB instance. In brief, for development, a self-signed
+certificate is usually sufficient, so the CLI tooling should have an option to
+automatically generate that for the user transparently. For production, we'd
+also like to make the certificate/key management easier and safe, options are
+discussed later.
+
+
+Internals
+^^^^^^^^^
+
+Under the neath, we'll have new files and folders in the EdgeDB well-known
+folder::
+
+    $WELLKNOWN/
+        +--config/
+        |    +--instance_ports.json
+        |    +--tls.json              <NEW>
+        |--credentials/
+        |    +--<instance>.json       <UPDATED>
+        |--certs/                     <NEW>
+        |    +--<instance>.cert.pem   <NEW>
+        |    +--<instance>.key.pem    <NEW>
+        ...
+
+The ``config/tls.json`` contains optional TLS settings for each instance::
+
+    {
+        "<instance>": {
+            "protocol_min": "<min TLS version>",
+            "protocol_max": "<max TLS version>",
+            "ciphers": "<available ciphers>"
+        },
+        ...
+    }
+
+If an instance has no record in ``config/tls.json``, the EdgeDB server options
+``--tls-protocol-min``, ``--tls-protocol-max`` and ``--tls-ciphers`` will be
+omitted and the default value will be used.
+
+``credentials/<instance>.json`` is updated to include the passphrase for the
+private key file, if any. For example::
+
+    {
+        "port": 10732,
+        "user": "edgedb",
+        "password": "....",
+        "database": "edgedb",
+        "tls_private_key_password": "...."
+    }
+
+Question: I don't feel comfortable storing the key passphrase here in clear
+text, especially if this is in production environment and we asked the user
+for passphrase but stored in this file. Perhaps we could use this JSON value
+only for the development process, and introduce a more secure (thus less
+convenient) way to obtain the private key password, like a customizable command
+as in PostgreSQL?
+
+Under the new folder ``certs/``, we'll place all certificates and keys used by
+the EdgeDB instances. The cert files will be named after the instance names.
+For development scenario we'll likely generate those certs for the user, while
+in production we'll "import" the certs specified by the user, by copying and
+renaming the cert and key files to the ``certs/`` folder.
```

Comment:
Does named cert file includes whole chain? Also this assumes that client certificates are used, right?

Reply:
> Does named cert file includes whole chain?

No I wasn't planning to include the root CA certificate. But the user should be responsible for imporing certfile with the necessary intermediate certificate in the chain for their client certificate verification to pass. So in other words, our client bindings trust only the dev CA root certificate (for dev only) in addition to the system CA repo, and the certificate verification on the client side depends on what's in the named cert file.

> Also this assumes that client certificates are used, right?

Sorry about the confusion! I didn't mean to involve client certificates. That's a good reminder though that we should also think about managing trusted client certificates. I'll add a new section for that, and the implementation may have a lower priority I guess.

Reply:
Oh, sorry I've misread this. So:

> in production we'll "import" the certs specified by the user, by copying and renaming the cert and key files to the ``certs/`` folder.

I don't think this makes sense. If user specified path to file on the command-line they assume that file is used, not copied somewhere. And if it copied, then when? (On each startup? Why whould you need copy in this case). Generally, let users manage private files, they will probably use their own private partitions for that.

Reply:
Also: client needs some way to discover root CA certs of the servers chain. This must either be a well-known location, or `credentials.json` itself.

Unless, we rely on universal root CAs on production, and use lets encrypt. Which this spec doesn't say we use. (And it's bad for some well-firewalled on-premise installations).